File: packages/cloud-agents-sdk/src/sse.ts
import type { StreamEvent } from "./types.ts";

export function parseSse(body: string): ReadonlyArray<StreamEvent> {
  const events: StreamEvent[] = [];
  for (const chunk of body.split("\n\n")) {
    if (chunk.trim().length === 0) continue;
    let id: string | undefined;
    let event = "message";
    let data = "";
    for (const line of chunk.split("\n")) {
      if (line.startsWith("id:")) id = line.slice(3).trim();
      else if (line.startsWith("event:")) event = line.slice(6).trim();
      else if (line.startsWith("data:")) data += line.slice(5).trim();
    }
    events.push({
      ...(id === undefined ? {} : { id }),
      event,
      data: data.length === 0 ? {} : (JSON.parse(data) as unknown),
    });
  }
  return events;
}

export function resumeSse(
  events: ReadonlyArray<StreamEvent>,
  lastEventId: string | undefined,
): ReadonlyArray<StreamEvent> {
  if (lastEventId === undefined) return events;
  const index = events.findIndex((event) => event.id === lastEventId);
  return index === -1 ? events : events.slice(index + 1);
}

File: packages/cloud-agents-sdk/python/t3_sdk/sse.py
from __future__ import annotations

import json


def parse_sse(body: str) -> list[dict]:
    events: list[dict] = []
    for chunk in body.split("\n\n"):
        if not chunk.strip():
            continue
        event = {"event": "message", "data": {}}
        data = ""
        for line in chunk.split("\n"):
            if line.startswith("id:"):
                event["id"] = line[3:].strip()
            elif line.startswith("event:"):
                event["event"] = line[6:].strip()
            elif line.startswith("data:"):
                data += line[5:].strip()
        if data:
            event["data"] = json.loads(data)
        events.append(event)
    return events


def resume_sse(events: list[dict], last_event_id: str | None) -> list[dict]:
    if last_event_id is None:
        return events
    for index, event in enumerate(events):
        if event.get("id") == last_event_id:
            return events[index + 1 :]
    return events
